File: dimdwarf-core/src/main/scala/net/orfjackal/dimdwarf/services/ServiceRunner.scala
package net.orfjackal.dimdwarf.services

import net.orfjackal.dimdwarf.mq.MessageReceiver
import org.slf4j.LoggerFactory

class ServiceRunner(service: Service, toService: MessageReceiver[Any]) extends Runnable {
  private val logger = LoggerFactory.getLogger(getClass)

  def run() {
    try {
      while (true) {
        val message = toService.take()
        logger.debug("PROCESS: {}", message)
        service.process(message)
      }
    } catch {
      case e =>
        logger.error("Internal error, service died", e)
        throw e
    }
  }
}

File: dimdwarf-core/src/main/java/net/orfjackal/dimdwarf/services/ServiceRunner.java
// Copyright © 2008-2010 Esko Luontola <www.orfjackal.net>
// This software is released under the Apache License 2.0.
// The license text is at http://dimdwarf.sourceforge.net/LICENSE

package net.orfjackal.dimdwarf.services;

import net.orfjackal.dimdwarf.context.*;

import javax.annotation.concurrent.ThreadSafe;
import java.util.concurrent.CountDownLatch;

@ThreadSafe
public class ServiceRunner implements Runnable {

    private final ServiceRegistration service;
    private final CountDownLatch started = new CountDownLatch(1);

    public ServiceRunner(ServiceRegistration service) {
        this.service = service;
    }

    public void awaitStarted() throws InterruptedException {
        started.await();
    }

    public void run() {
        Context context = service.getContext().get();
        ThreadContext.runInContext(context, new Runnable() {
            public void run() {
                ServiceRunnable sr = service.getService().get();
                sr.start();
                started.countDown();
                sr.run();
            }
        });
    }
}
